feat: Add tool for extracting annotions of species in an ODE model

Rakesh-Seenu opened the pull request:
# For authors

https://github.com/user-attachments/assets/98cc41fa-9315-4296-8c5c-dbe70c900a3a

## Description

This PR introduces a new tool in Talk2BioModels that helps retrieve species annotations from a model. 
When the tool is prompted **tool uses get_miriam_annotation** function from basico library in the backend  
to retrieve the annotations of the specified species.

When this function is called, it returns the following details:

- Name – The species name as recorded in the database.
- URL – A link to an external database where the species is listed.
- Qualifier – A category or tag that describes the species.

However, the species name is often not very informative, Hence I have developed modules that make API calls to the external databases to fetch the descriptions of the species name. 
Since different species are listed in different databases, fetching their descriptions requires extra steps.

### How the Tool Fetches Descriptions

The species URLs point to multiple external databases, where more detailed descriptions are available. 
I have implemented three API handler files to retrieve description from specific set of databases using the provided URL.

When a user asks for a species annotation, the tool first retrieves the basic details (URL, Name, Qualifier).
Then, based on which database the species belongs to, the tool calls the appropriate API handler file to fetch its description.
Finally, all the information is displayed in an easy-to-read table.

### Results Are Shown to the User is in the demo

The retrieved annotations are displayed in a scrollable table with the following columns:

* Species Name   : The name of the species.
* Description	    : A brief explanation of the species from the database.
* Database	    : The name of the database where the species is listed.
* ID	                    : A unique identifier, shown as a clickable hyperlink that directs the user to the species database page.
* Qualifier	    : Additional information about the species annotation.

### What This Tool Can Do :

✅ Retrieve annotations for one or multiple species in a model.
✅ Fetch all species annotations in a given model.
✅ Remember the model ID from chat history, so users don’t need to enter it each time.
✅ Handle errors gracefully – If a species is not found or its description is missing, the user is notified in the front end.

### Upcoming Feature Enhancements
This version only allows users to view species annotations, but future updates will add more features:

- Adding more Databases for API Calling
    Include Databases like interpro and go
- Editing & Updating Annotations
    Users will be able to edit and update species annotations directly in the table.
- Support for Abstract Questions
    Instead of asking for specific species by name, users will be able to make broader requests, such as:
    "Show me annotations of all the InterLeukins in the model ."

## Fixes #57  (issue)

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test ( test_get_annotaion ) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

dmccloskey commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -0,0 +1,288 @@
+        """
+        Process link to format it correctly.
+        """
+        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/"]
```
I am a bit concerned this will be difficult to maintain as there are a LOT of different ontologies. What is the problem that this method solves in regard to link formatting? If it is needed, is it possible to make it more general? Another idea would be just to include all of the ontology abbreviations from OLS if it is the same for each of them.

Rakesh-Seenu replied:
Problem faced is that, in Some cases the link don't work when get_miriam_annotation is called because they have database name in between the Link . 

For example in model 537 for species sR

sR  http://identifiers.org/pato/PATO:0001537   

As the returned link is invalid, to make it a valid link i have use these substrings to remove unnecessary part in link.	
yes we can include all of ontology terms, May be in next release as i don't know all the ontology abbreviations

dmccloskey replied:
I see. I remember running into this issue as well once...

All of the OLS abbreviations can be found here: https://www.ebi.ac.uk/ols4/ontologies.

Rakesh-Seenu replied:
Thanks for the link

dmccloskey commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -0,0 +1,288 @@
+        for database, identifiers in grouped_data.items():
+            if database == 'uniprot':
+                results.update(search_uniprot_labels(identifiers))
+            elif database in {'pato', 'chebi', 'sbo', 'fma', 'pr'}:
```
I would recommend creating a constant variable at the beginning of the module that has the list of OLS ontology abbreviations so that at the very least it only needs to be updated in one place.

Rakesh-Seenu replied:
Resolved by creating the single variable for list of ontology abbreviations called "ols_ontology_abbreviations"

dmccloskey reviewed (changes requested):
Nice work 💪. I am very pleased to see the thought that went into fetching additional information from different ontologies and databases. @awmulyadi I think the APIs could be useful for you for the enrichment part. At least chemicals, proteins, genes, GO, and diseases appear to be covered as many are included in OLS.

I have several comments in regard to the testing and how the multiple OLS sub databases are handled. Please reach out if you have any questions.

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -0,0 +1,288 @@
+        # If empty, return a message
+        if annotations_df.empty:
+            logger.warning("The annotations dataframe is empty.")
+            return f'''The following species do not exist:
```
yes, I have replaced it with prepare_content_msg function

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -0,0 +1,288 @@
+        logger.info("Running the GetAnnotationTool tool for species %s", list_species_names)
+
+        # Prepare the model object
+        sbml_file_path = state['sbml_file_path'][-1] if state['sbml_file_path'] else None
```
I have added new test case for the no model, in function test_all_species()

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -0,0 +1,288 @@
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+def prepare_content_msg(species_not_found: List[str],
```
Thanks :)

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tests/test_get_annotation.py:
```diff
@@ -0,0 +1,155 @@
+# In the following test cases, we will test the API functions,
+# which were not tested in the cases above.
+
+def test_search_uniprot_labels():
```
yes, I have removed the repeated code in test_get_annotation

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tests/test_api.py:
```diff
@@ -0,0 +1,51 @@
+    '''
+    Test the fetch_from_ols function.
+    '''
+    term = "GO:ABC123"
```
I have added positive case also for ols function

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/api/kegg.py:
```diff
@@ -0,0 +1,83 @@
+
+def fetch_kegg_names(ids: List[str], batch_size: int = 10) -> Dict[str, str]:
+    """
+    Fetch the names of multiple KEGG entries using the KEGG REST API.
```
updated as suggested

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/api/kegg.py:
```diff
@@ -0,0 +1,83 @@
+    Fetch the names of multiple KEGG entries using the KEGG REST API.
+
+    Args:
+        ids (List[str]): List of KEGG compound IDs.
```
It will work for all the kegg abbrivations, I have removed compound in Args

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tests/test_get_annotation.py:
```diff
@@ -0,0 +1,155 @@
+                config=config
+            )
+    current_state = app.get_state(config)
+    print (current_state.values["dic_annotations_data"])
```
No, I have just added those for testing purpose. I have removed those now.

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tests/test_get_annotation.py:
```diff
@@ -0,0 +1,155 @@
+    # Loop through the reversed messages until a
+    # ToolMessage is found.
+    artifact_was_none = False
+    for msg in reversed_messages:
```
I have now put comment to explain working of those valid and invalid species.

Rakesh-Seenu commented on aiagents4pharma/talk2biomodels/tests/test_get_annotation.py:
```diff
@@ -0,0 +1,155 @@
+        current_state = app.get_state(config)
+        dic_annotations_data = current_state.values["dic_annotations_data"]
+        print (dic_annotations_data)
+        assert isinstance(dic_annotations_data, list)
```
Updated the function of test_all_species which covers all the the expected outputs

dmccloskey replied:
Cool. However, I am still missing the test for `prepare_content_msg` and the comparison of the expected string produced by this method for all species (unless I somehow missed it).

Rakesh-Seenu replied:
reversed_messages = current_state.values["messages"][::-1]

    # Covered all the use case for the expecetd sting on all the species
    test_condition = False
    for msg in reversed_messages:
        if isinstance(msg, ToolMessage) and msg.name == "get_annotation":
            print("ToolMessage Content:", msg.content)  # Debugging output
            if msg.artifact is None and ('ORI' in msg.content or
                                         "Successfully extracted annotations for the species"
                                         in msg.content or "Error" in msg.content):
                test_condition = True
                break

    dic_annotations_data = current_state.values["dic_annotations_data"]

    assert isinstance(dic_annotations_data, list),\
        f"Expected a list for model {model_id}, got {type(dic_annotations_data)}"
    assert len(dic_annotations_data) > 0,\
        f"Expected species data for model {model_id}, but got empty list"
    assert test_condition # Expected output is validated
    
 Rather than using prepare_content_msg, I have added this code where it checks the expected and the tool message produced.

dmccloskey replied:
I see. 

I think this can be made much clearer with my suggestions above.

Rakesh-Seenu replied:
I have created new function  test_prepare_content_msg() for checking expected messages

dmccloskey commented on aiagents4pharma/talk2biomodels/tests/test_api.py:
```diff
@@ -19,9 +20,14 @@ def test_fetch_from_ols():
-    term = "GO:ABC123"
-    label = fetch_from_ols(term)
-    assert label.startswith("Error: 404")
+    term_1 = "GO:0005886" #Negative result
```
```suggestion
    term_1 = "GO:0005886" #Positive result
```

Rakesh-Seenu replied:
I have corrected the comment

dmccloskey commented on aiagents4pharma/talk2biomodels/tests/test_api.py:
```diff
@@ -19,9 +20,14 @@ def test_fetch_from_ols():
-    label = fetch_from_ols(term)
-    assert label.startswith("Error: 404")
+    term_1 = "GO:0005886" #Negative result
+    term_2 = "GO:ABC123" #Positive result
```
```suggestion
    term_2 = "GO:ABC123" #Negative result
```

Rakesh-Seenu replied:
I have corrected the comment

dmccloskey commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -266,10 +261,13 @@ def _fetch_descriptions(self, data: List[dict[str, str]]) -> dict[str, str]:
         # In the following loop, we fetch the descriptions for the identifiers
         # based on the database type.
+        # Constants
+        ols_ontology_abbreviations = {'pato', 'chebi', 'sbo', 'fma', 'pr'}
```
Much cleaner with a named variable for the abbreviations 🙂. It looks like there are two places where this list is used. Would it be possible to make this a const global variable at the top of the file or in the __init__ so that there is no replication?

Rakesh-Seenu replied:
I have made it as a constant global variable

dmccloskey reviewed (changes requested):
The review updates are looking much better 👍. Just a few minor comments at this point.

dmccloskey commented on aiagents4pharma/talk2biomodels/tests/test_get_annotation.py:
```diff
@@ -0,0 +1,154 @@
+
+    reversed_messages = current_state.values["messages"][::-1]
+
+    # Covered all the use case for the expecetd sting on all the species
```
```suggestion
    # Covers all of the use cases for the expected string on all the species
```

Rakesh-Seenu replied:
I have updated the comment

Rakesh-Seenu commented:
Hi, @dmccloskey making some more updates I will let you know, Could you please merge the pull request after making some more changes.

dmccloskey reviewed (approved):
The updates look good 👍. Nice work 💪.

Before we can merge, please take care of the following:

1. the merge conflicts
2. open an issue for adding all of the OLS abbreviations

Rakesh-Seenu commented:
Hi @dmccloskey, I am still working on this PR on some updates , Please review the code only after you have heard from me. Thx

Rakesh-Seenu commented:
> The updates look good 👍. Nice work 💪.
> 
> Before we can merge, please take care of the following:
> 
> 1. the merge conflicts
> 2. open an issue for adding all of the OLS abbreviations

Hi @dmccloskey,

Thank you for your feedback. I’ve addressed the items you mentioned and would like to provide a detailed update on the recent changes:

- Handling the iOS Species Return Issue:
During testing, I discovered that iOS was returning None when it should have been returning the species. I investigated the issue, implemented a fix to properly handle the case when the species needs to be returned, and added a new test case specifically for model 20. This update ensures that the behavior is now consistent and the error has been resolved.

In get_annotaion I have added below code : 
![image](https://github.com/user-attachments/assets/1c40fa3d-d195-4f4d-8d8b-5f1533c4043f)

In test_get_annotaion I have updated as follows :
![image](https://github.com/user-attachments/assets/278afb39-a8bd-4c9c-b105-090b684f3417)

- Updating OLS Abbreviations for Model like 56:
I noticed that the [go data base] abbreviation was missing in the OLS for model 56. I’ve now added this abbreviation, and as demonstrated in the attached screenshot, it is displaying correctly. 
![image](https://github.com/user-attachments/assets/24defee7-d31b-4f97-b474-88123632017d)

In addition, I will open a separate issue to track the addition of all missing OLS abbreviations to ensure comprehensive coverage across models.

**Additional Verifications:**
I have run both pylint and coverage tests. The code meets our style guidelines, and the test coverage is complete with no conflicts, ensuring that all updates are in line with our quality standards.
![image](https://github.com/user-attachments/assets/2851ecdd-0955-4376-aec5-efe81b529b23)
![image](https://github.com/user-attachments/assets/0f7b31b4-06e7-4a5c-8353-245abd0e967f)

Please let me know if you have any further questions or require additional changes :)

dmccloskey commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -229,7 +234,7 @@ def _process_link(self, link: str) -> str:
         Process link to format it correctly.
         """
-        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/"]
+        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/", "go/"]
```
```suggestion
```

dmccloskey commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -229,7 +234,7 @@ def _process_link(self, link: str) -> str:
         """
-        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/"]
+        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/", "go/"]
         for substring in substrings:
```
```suggestion
        for substring in ols_ontology_abbreviations:
```

dmccloskey commented on aiagents4pharma/talk2biomodels/tools/get_annotation.py:
```diff
@@ -229,7 +234,7 @@ def _process_link(self, link: str) -> str:
-        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/"]
+        substrings = ["chebi/", "pato/", "pr/", "fma/", "sbo/", "go/"]
         for substring in substrings:
             if substring in link:
```
```suggestion
            if substring + '/' in link:
```

dmccloskey reviewed (approved):
Please integrate my suggestions (you may have to modify slightly), check that the tests and linting pass, and then merge 🙂.

dmccloskey commented:
@Rakesh-Seenu It looks like there are a few linting failures and some failing tests. Please take a look and ping me when they have been resolved.
